server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1911,6 +1935,81 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.nextToken(), parser);
             return builder.build();
         }
+
+        private IndexMetadata reuseMappings(IndexMetadata indexMetadata) {
+            if (indexMetadata.mapping() == null) {
+                return indexMetadata;
+            }
+
+            String digest = indexMetadata.mapping().getDigest();
+            Entry entry = cache.get(digest);
+            if (entry != null) {
+                cache.put(digest, new Entry(entry.refCounter + 1, entry.mapping));
+                IndexMetadata.Builder imBuilder = new IndexMetadata.Builder(indexMetadata);
+                imBuilder.putMapping(entry.mapping);
+                return imBuilder.build();
+            } else {
+                entry = new Entry(1, indexMetadata.mapping());
+                cache.put(digest, entry);
+                return indexMetadata;
+            }
+        }
+
+        private void reuseMappings(IndexMetadata.Builder indexMetadataBuilder) {
+            if (indexMetadataBuilder.mapping() == null) {
+                return;
+            }
+
+            String digest = indexMetadataBuilder.mapping().getDigest();
+            Entry entry = cache.get(digest);
+            if (entry != null) {
+                cache.put(digest, new Entry(entry.refCounter + 1, entry.mapping));
+                indexMetadataBuilder.putMapping(entry.mapping);
+            } else {
+                entry = new Entry(1, indexMetadataBuilder.mapping());
+                cache.put(digest, entry);
+            }
+        }
+
+        private void cleanupUnusedEntry(IndexMetadata previous) {
```

Comment:
Because the cache is passed down to new `Metadata` instances, a cleanup mechanism is needed.
This approach with counting the times a `MappignMetata` instance is used and removing entries when there are no usages seems cheaper compared to checking each metadata is build whether all entries are used and then removed the unused entries. But this approach is also more complex. I'm also okay to change the cleanup logic to be simpler and checking each cache entry whether it is used in any of the indices and if unused remove.

Reply:
If the simpler version doesn't come with problematic runtime overhead it might be fine? It seems like it should be cheap now that we can cheaply compare mappings?

Reply:
I think that is true. Also I find this implementation a bit fragile. I will change the cleanup implementation to just check for each cache entry whether it is used by any IndexMetadata.

Reply:
Implemented: 15b05afb51274069320f1db6ae29f62b95e91348